From: Kefeng Wang <wangkefeng.wang@huawei.com>
Subject: mm: hugetlb: optimize replace_free_hugepage_folios()
Date: Wed, 14 Jan 2026 21:55:12 +0800

Turn back to use alloc_and_dissolve_hugetlb_folio() since Oscar pointed
out that the return value is different.  Then adding gigantic folio check
independently, also update changelog.

Link: https://lkml.kernel.org/r/20260114135512.2159799-1-wangkefeng.wang@huawei.com
Signed-off-by: Kefeng Wang <wangkefeng.wang@huawei.com>
Cc: Brendan Jackman <jackmanb@google.com>
Cc: David Hildenbrand <david@kernel.org>
Cc: Jane Chu <jane.chu@oracle.com>
Cc: Johannes Weiner <hannes@cmpxchg.org>
Cc: Matthew Wilcox (Oracle) <willy@infradead.org>
Cc: Muchun Song <muchun.song@linux.dev>
Cc: Oscar Salvador <osalvador@suse.de>
Cc: Sidhartha Kumar <sidhartha.kumar@oracle.com>
Cc: Vlastimil Babka <vbabka@suse.cz>
Cc: Zi Yan <ziy@nvidia.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 mm/hugetlb.c |   10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

--- a/mm/hugetlb.c~mm-hugetlb-optimize-replace_free_hugepage_folios-v2
+++ a/mm/hugetlb.c
@@ -2810,6 +2810,7 @@ int replace_free_hugepage_folios(unsigne
 	struct page *page;
 	struct hstate *h;
 	LIST_HEAD(list);
+	int ret = 0;
 
 	/* Avoid pfn iterations if no free non-gigantic huge pages */
 	for_each_hstate(h) {
@@ -2835,9 +2836,13 @@ int replace_free_hugepage_folios(unsigne
 
 			/* Not to disrupt normal path by vainly holding hugetlb_lock */
 			if (folio_test_hugetlb(folio) && !folio_ref_count(folio)) {
-				if (isolate_or_dissolve_huge_folio(folio, &list))
+				if (order_is_gigantic(folio_order(folio)))
 					return -ENOMEM;
 
+				ret = alloc_and_dissolve_hugetlb_folio(folio, &list);
+				if (ret)
+					break;
+
 				putback_movable_pages(&list);
 			}
 		} else if (PageBuddy(page)) {
@@ -2851,11 +2856,10 @@ int replace_free_hugepage_folios(unsigne
 			if (order <= MAX_PAGE_ORDER)
 				nr = 1UL << order;
 		}
-
 		start_pfn += nr;
 	}
 
-	return 0;
+	return ret;
 }
 
 void wait_for_freed_hugetlb_folios(void)
_
